predicators/envs/gridworld.py

```diff
@@ -0,0 +1,647 @@
+                                self._Clear_holds)
+        self._GoalHack = Predicate("GoalHack", [
+            self._bottom_bun_type, self._patty_type, self._cheese_type,
+            self._tomato_type, self._top_bun_type
+        ], self._GoalHack_holds)
+
+        # Static objects (exist no matter the settings)
+        self._robot = Object("robby", self._robot_type)
+        self._grill = Object("grill", self._grill_type)
+        self._cutting_board = Object("cutting_board", self._cutting_board_type)
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "gridworld"
+
+    @property
+    def types(self) -> Set[Type]:
+        return {
+            self._object_type, self._item_type, self._station_type,
+            self._robot_type, self._patty_type, self._tomato_type,
+            self._cheese_type, self._bottom_bun_type, self._top_bun_type,
+            self._grill_type, self._cutting_board_type
+        }
+
+    def _get_tasks(self, num: int,
+                   rng: np.random.Generator) -> List[EnvironmentTask]:
+        del rng  # unused
+        tasks = []
+
+        # Add robot, grill, and cutting board
+        state_dict = {}
+        hidden_state = {}
+        state_dict[self._robot] = {
+            "row": 2,
+            "col": 2,
+            "fingers": 0.0,
+            "dir": 3
+        }
+        state_dict[self._grill] = {"row": 2, "col": 3, "z": 0}
+        state_dict[self._cutting_board] = {"row": 1, "col": 3, "z": 0}
+
+        # Add patty
+        patty = Object("patty", self._patty_type)
+        state_dict[patty] = {"row": 0, "col": 0, "z": 0}
+        hidden_state[patty] = {"is_cooked": 0.0, "is_held": 0.0}
+
+        # Add tomato
+        tomato = Object("tomato", self._tomato_type)
+        state_dict[tomato] = {"row": 0, "col": 1, "z": 0}
+        hidden_state[tomato] = {"is_sliced": 0.0, "is_held": 0.0}
+
+        # Add cheese
+        cheese = Object("cheese", self._cheese_type)
+        state_dict[cheese] = {"row": 3, "col": 0, "z": 0}
+        hidden_state[cheese] = {"is_held": 0.0}
+
+        # Add top bun
+        top_bun = Object("top_bun", self._top_bun_type)
+        state_dict[top_bun] = {"row": 3, "col": 1, "z": 0}
+        hidden_state[top_bun] = {"is_held": 0.0}
+
+        # Add bottom bun
+        bottom_bun = Object("bottom_bun", self._bottom_bun_type)
+        state_dict[bottom_bun] = {"row": 0, "col": 2, "z": 0}
+        hidden_state[bottom_bun] = {"is_held": 0.0}
+
+        goal = {
+            GroundAtom(self._On, [patty, bottom_bun]),
+            GroundAtom(self._On, [cheese, patty]),
+            GroundAtom(self._On, [tomato, cheese]),
+            GroundAtom(self._On, [top_bun, tomato]),
+            GroundAtom(self._IsCooked, [patty]),
+            GroundAtom(self._IsSliced, [tomato]),
+        }
+
+        for _ in range(num):
+            state = utils.create_state_from_dict(state_dict)
+            state.simulator_state = hidden_state
```

Comment:
Note that the VLM predicate invention pipeline currently uses the simulator state to store renderings of the observation to pass to the VLM for predicate invention purposes. We can certainly modify this, and I don't think it's important for this specific PR, but we should probably chat about this and implement something that allows VLM stuff to be run on this env in the next PR

Reply:
I think it's fine to leave in there, since we only need to read things from the simulator state dictionary for a specific key. At the moment, I'm not comparing `compare_fn(state1.simualtor_state, state2.simulator_state)` anywhere. 

^ I would like to add this comparison possibility -- in which case maybe we can have a flag set the specific type of simulator state comparison function that is used (or better -- maybe we make a new `State.allclose2()` function that also takes in a comparison function as an argument for how to compare the simulator state. 

++ on dealing with that in the next PR